commands/list_keypairs.go

```diff
@@ -23,84 +27,208 @@ var (
+			subtext = "Please specify which cluster to list key pairs for, by using the '-c' or '--cluster' argument."
+		default:
+			headline = err.Error()
+		}
+
+		fmt.Println(color.RedString(headline))
+		if subtext != "" {
+			fmt.Println(subtext)
+		}
+		os.Exit(1)
 	}
-	if cmdClusterID == "" {
+}
+
+// listKeypairsValidate validates our pre-conditions and returns an error in
+// case something is missing.
+// If no clusterID argument is given, and a default cluster can be determined,
+// the listKeypairsArguments given as argument will be modified to contain
+// the clusterID field.
+func listKeypairsValidate(args *listKeypairsArguments) error {
+	if config.Config.Token == "" && args.token == "" {
+		return errors.New(errNotLoggedIn)
+	}
+	if args.clusterID == "" {
 		// use default cluster if possible
 		clusterID, _ := config.GetDefaultCluster(requestIDHeader, listKeypairsActivityName, cmdLine, cmdAPIEndpoint)
 		if clusterID != "" {
 			cmdClusterID = clusterID
 		} else {
-			return errors.New("No cluster given. Please use the -c/--cluster flag to set a cluster ID.")
+			return errors.New(errClusterIDNotSpecified)
 		}
 	}
+
 	return nil
 }
 
-// listKeypairs is the function called to list keypairs and display
+// listKeypairsOutput is the function called to list keypairs and display
 // errors in case they happen
-func listKeypairs(cmd *cobra.Command, args []string) {
-	output, err := keypairsTable()
+func listKeypairsOutput(cmd *cobra.Command, extraArgs []string) {
+	args := defaultListKeypairsArguments()
+	result, err := listKeypairs(args)
+
+	// error output
 	if err != nil {
-		fmt.Println(color.RedString("Error: %s", err))
-		if _, ok := err.(APIError); ok {
-			dumpAPIResponse((err).(APIError).APIResponse)
+		var headline string
+		var subtext string
+
+		switch err.Error() {
+		case errNotLoggedIn:
+			headline = "You are not logged in."
+			subtext = "Please log in using 'gsctl login <email>' or set an auth token as a command line argument."
+			subtext += " See `gsctl list keypairs --help` for details."
+		case errClusterIDNotSpecified:
+			headline = "No cluster ID specified."
+			subtext = "Please specify which cluster to list key pairs for, by using the '-c' or '--cluster' argument."
+		case errNotAuthorized:
+			headline = "You are not authorized for this cluster."
+			subtext = "You have no permission to access key pairs for this cluster. Please check your credentials."
+		case errClusterNotFound:
+			headline = "The cluster does not exist."
+			subtext = fmt.Sprintf("We couldn't find a cluster with the ID '%s' via API endpoint %s.", args.clusterID, args.apiEndpoint)
+		case errInternalServerError:
+			headline = "An internal error occurred."
+			subtext = "Please notify the Giant Swarm support team, or try listing key pairs again in a few moments."
+		default:
+			headline = err.Error()
+		}
+
+		fmt.Println(color.RedString(headline))
+		if subtext != "" {
+			fmt.Println(subtext)
 		}
 		os.Exit(1)
```

Comment:
I've been changing the code lately to be separate the business logic from the output printing. That at least allows for testing of the business logic, and that works quite well now.

So far we are not testing the printed output in unit tests. Might be that we start doing that at some point.